BigQuery/tests/System/RegionalizationTest.php

```diff
@@ -0,0 +1,207 @@
+            // we get a 404.
+        }
+
+        $this->assertExceptionShouldBeThrown($location, $caught);
+    }
+
+    /**
+     * @dataProvider locationDataProvider
+     */
+    public function testRunExtractJob($location)
+    {
+        $caught = false;
+        $extract = self::$tableAsia->extract(
+            self::$bucket->object(self::TESTING_PREFIX)
+        )->location($location);
+
+        try {
+            self::$tableAsia->startJob($extract);
+        } catch (NotFoundException $e) {
+            $caught = true;
+        }
+
+        $this->assertExceptionShouldBeThrown($location, $caught);
+    }
+
+    /**
+     * @dataProvider locationDataProvider
+     */
+    public function testRunLoadJob($location)
+    {
+        $caught = false;
+        $load = self::$tableAsia->load(
+            file_get_contents(__DIR__ . '/data/table-data.json')
+        )
+            ->sourceFormat('NEWLINE_DELIMITED_JSON')
+            ->location($location);
+
+        try {
+            self::$tableAsia->startJob($load);
+        }catch (ServiceException $e) {
+            $code = $e->getCode();
+            if ($code === 403 || $code === 404) {
+                $caught = true;
+            }
+        }
+
+        $this->assertExceptionShouldBeThrown($location, $caught);
+    }
+
+    /**
+     * @dataProvider locationDataProvider
+     */
+    public function testGetsJob($location)
+    {
+        $caught = false;
+        $query = self::$client->query(
+            sprintf(
+                self::QUERY_TEMPLATE,
+                self::$datasetAsia->id(),
+                self::$tableAsia->id()
+            )
+        )->location(self::REGION_ASIA);
+        $job = self::$client->startQuery($query);
+
+        try {
+            self::$client->job($job->id(), [
+                'location' => $location
+            ])->reload();
+        }catch (NotFoundException $e) {
+            $caught = true;
+        }
+
+        $this->assertExceptionShouldBeThrown($location, $caught);
+    }
+
+    /**
+     * @dataProvider locationDataProvider
+     */
+    public function testCancelsJob($location)
+    {
+        $caught = false;
+        $query = self::$client->query(
+            sprintf(
+                self::QUERY_TEMPLATE,
+                self::$datasetAsia->id(),
+                self::$tableAsia->id()
+            )
+        )->location(self::REGION_ASIA);
+        $job = self::$client->startQuery($query);
+
+        try {
+            self::$client->job($job->id(), [
+                'location' => $location
+            ])->cancel();
+        }catch (NotFoundException $e) {
+            $caught = true;
+        }
+
+        $this->assertExceptionShouldBeThrown($location, $caught);
+    }
+
+    public function locationDataProvider()
+    {
+        return [
+            [self::REGION_ASIA],
+            [self::REGION_US]
+        ];
+    }
+
+    private function assertExceptionShouldBeThrown($location, $caught)
```

Comment:
I'd prefer not to hide this logic in a helper function. It should be inlined. In fact, I'd actually prefer if we had a little extra duplicate code in the tests and didn't catch the exception at all when we don't expect one to be thrown.

Reply:
Just to be sure I am clear, would you like to have the dataProvider removed, the assert function, or both?

Reply:
I was thinking both, but if you can think of a way to use the dataProvider without having an if-statement to check the location, then go for it. I'd like for it to be crystal clear what the logic is in the test.